File: java/vm/src/main/java/simplevm/vm/VirtualMachine.java
package simplevm.vm;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

import static simplevm.vm.Bytecode.*;

public class VirtualMachine {

    public static class Exception extends java.lang.RuntimeException {
        public Exception(String message) {
            super(message);
        }
    }

    // Tracing
    //
    boolean trace = false;
    private void trace(String message) {
        if (trace) {
            System.out.println("(IP " + ip + "):" + message);
        }
    }

    // Dump
    //
    private void dump() {
        System.out.println("SimpleVM DUMP");
        System.out.println("=============");
        /*
        System.out.println("IP: " + ip);
        System.out.println("Working stack (SP " + sp + "): " + Arrays.toString(Arrays.copyOfRange(stack, 0, sp+1)));
        System.out.println("Globals: " + Arrays.toString(globals));
        System.out.println("Call stack: ");
        for (int f = frames.size(); f != 0; f--) {
            CallFrame cf = frames.get(f - 1);
            System.out.println("  Call Frame " + (f - 1) + ":");
            System.out.println("  +-- Return Address: " + cf.returnAddress);
            System.out.println("  +-- Locals: " + Arrays.toString(cf.locals));
        }
        */
    }

    public void execute(int opcode, int... operands) {

    }
    public void execute(int[] code) {

    }
}

File: cs/vm/VirtualMachine.cs
namespace vm;

public enum Bytecode 
{
    NOP,
    DUMP,
    TRACE,
    PRINT,
    HALT,
    FATAL,

    // Stack opcodes
    CONST,
    POP,

    // Math opcodes (binary)
    ADD,
    SUB,
    MUL,
    DIV,
    MOD,

    // Math opcodes (unary)
    ABS,
    NEG,

    // Comparison
    EQ,
    NEQ,
    GT,
    LT,
    GTE,
    LTE,

    // Branching opcodes
    JMP,
    JMPI,
    RJMP,
    RJMPI,
    JZ,
    JNZ,

    // Globals
    GSTORE,
    GLOAD,

    // Procedures/locals
    CALL,
    RET,
    LOAD,
    STORE
}


public class VirtualMachine
{
    // Tracing
    //
    bool trace = false;
    private void Trace(string message)
    {
        if (trace)
            Console.WriteLine("TRACE: {0}", message);
    }

    // Diagnostics
    //
    private void Dump()
    {
        Console.WriteLine("SimpleVM - DUMP");
        Console.WriteLine("===============");
        // Uncomment when you implement Execute(Bytecode[])
        /*
        Console.WriteLine("IP: {0} / Trace: {1}", IP, trace);
        Console.WriteLine("Working stack (SP {0}): {1}", SP, String.Join(", ", Stack));
        */
        // Uncomment when you implement global values
        /*
        Console.WriteLine("Globals: {0}", Globals);
        */
        // Uncomment when you implement CallFrames (procedures)
        /*
        Console.WriteLine("Call stack: ");
        for (int f = Frames.Count - 1; f > -1; f--) {
            CallFrame cf = Frames[f];
            Console.WriteLine("  Call Frame {0}:", f);
            Console.WriteLine("  +- ReturnAddr: {0}", cf.ReturnAddress);
            Console.WriteLine("  +- Locals: {0}", cf.Locals);
        }
        */
    }

    public void Execute(Bytecode opcode, params int operands)
    {

    }
    public void Execute(Bytecode[] code)
    {
        
    }
}
